src/Features/VisualBasic/Portable/ConvertForEachToFor/VisualBasicConvertForEachToForCodeRefactoringProvider.vb

```diff
@@ -0,0 +1,185 @@
+                    DirectCast(generator.LiteralExpression(0), ExpressionSyntax),
+                    DirectCast(generator.SubtractExpression(
+                        generator.MemberAccessExpression(
+                            collectionVariable, foreachInfo.CountName), generator.LiteralExpression(1)), ExpressionSyntax)),
+                bodyStatement,
+                nextStatement)
+
+            If foreachInfo.RequireCollectionStatement Then
+                ' this is to remove blank line between newly added collection statement with "For" keyword.
+                ' default VB formatting rule around elastic trivia and end of line trivia is converting elastic trivia
+                ' to end of line trivia causing there to be 2 line breaks. this fix that issue. not changing the default
+                ' rule since it will affect other ones that having opposite desire.
+                forBlock = forBlock.WithLeadingTrivia(SyntaxFactory.TriviaList())
+            Else
+                ' transfer comment on "For Each" to "For"
+                forBlock = forBlock.WithLeadingTrivia(forEachBlock.GetLeadingTrivia())
+            End If
+
+            ' transfer comment at then end of "For Each" to "For"
+            forBlock = forBlock.WithForStatement(forBlock.ForStatement.WithTrailingTrivia(forEachBlock.ForEachStatement.GetLastToken().TrailingTrivia))
+
+            editor.ReplaceNode(forEachBlock, forBlock)
+        End Sub
+
+        Private Function GetForLoopBody(
+            generator As SyntaxGenerator, foreachInfo As ForEachInfo,
+            collectionVariableName As SyntaxNode, indexVariable As SyntaxToken) As SyntaxList(Of StatementSyntax)
+
+            Dim forEachBlock = foreachInfo.ForEachStatement
+            If forEachBlock.Statements.Count = 0 Then
+                Return forEachBlock.Statements
+            End If
+
+            Dim foreachVariable As SyntaxNode = Nothing
+            Dim type As SyntaxNode = Nothing
+            GetVariableNameAndType(forEachBlock.ForEachStatement, foreachVariable, type)
+
+            ' use original text
+            Dim foreachVariableToken = generator.Identifier(foreachVariable.ToString())
+
+            ' create varialbe statement
+            Dim variableStatement = AddItemVariableDeclaration(
+                generator, type, foreachVariableToken, foreachInfo.ForEachElementType, collectionVariableName, indexVariable)
+
+            Return forEachBlock.Statements.Insert(0, DirectCast(variableStatement, StatementSyntax))
+        End Function
+
+        Private Sub GetVariableNameAndType(
+            forEachStatement As ForEachStatementSyntax, ByRef foreachVariable As SyntaxNode, ByRef type As SyntaxNode)
+
+            Dim controlVariable = forEachStatement.ControlVariable
+
+            Dim declarator = TryCast(controlVariable, VariableDeclaratorSyntax)
+            If declarator IsNot Nothing Then
+                foreachVariable = declarator.Names(0)
```

Comment:
do we know that names is always non-empty?

Reply:
checked with IOperation. I believe I also have test for in in C#. anyway, in future iteration, we probably will allow code tranformation even if there is no variable here.

Reply:
> checked with IOperation. 

Sorry... can clarify what this means in this context?  How does IOperation let you know that this list of names won't be empty (either currently, or in the future)?

>  I believe I also have test for in in C#. anyway

Sorry, i'm not sure how that is applicable to this code.  This is you indexing directly into the VB syntax model.  It is guaranteed that a VB declarator always has at least one name?

> we probably will allow code tranformation even if there is no variable here.

Sure.  my concern is not about allowing or not.  My concern is if this will crash or not due to indexing out of bounds?

Reply:
I checked that IForLoop having 1 locals. currently it won't offer refactoring if there is missing variable.